### PR TITLE
Exiting SSE transport receive loop when transport is stopping

### PR DESCRIPTION
**Pull request by moozzyk**
Fixes: #650

### CODE REVIEW

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 1]
```diff
@@ -116,6 +116,7 @@ public Task StartAsync(Uri url, Channel<byte[], SendMessage> application)
             {
                 _transportCts.Cancel();
                 stream.Dispose();
+                readCancellationRegistration.Dispose();
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: this should be disposed before canceling transportCts

**Comment by moozzyk** [comment]
CI is failing:

[xUnit.net 00:00:01.3510785]       System.InvalidOperationException : No authenticationScheme was specified, and there was no DefaultChallengeScheme found.
[xUnit.net 00:00:01.3520826]       Stack Trace:
[xUnit.net 00:00:01.3532656]            at Microsoft.AspNetCore.Authentication.AuthenticationService.<ChallengeAsync>d__11.MoveNext()
[xUnit.net 00:00:01.3533462]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:00:01.3534042]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:00:01.3534655]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:00:01.3535552]         C:\projects\signalr\src\Microsoft.AspNetCore.Sockets.Http\Internal\AuthorizeHelper.cs(50,0): at Microsoft.AspNetCore.Sockets.Internal.AuthorizeHelper.<AuthorizeAsync>d__0.MoveNext()

I am seeing the same failures locally...

**Comment by BrennanConroy** [comment]
Should be good now

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 2]
```diff
@@ -71,25 +71,25 @@ public Task StartAsync(Uri url, Channel<byte[], SendMessage> application)
             var response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
 
             var stream = await response.Content.ReadAsStreamAsync();
-
-            var pipelineReader = stream.AsPipelineReader();
+            var pipelineReader = stream.AsPipelineReader(cancellationToken);
+            var readCancellationRegistration = cancellationToken.Register(() => pipelineReader.CancelPendingRead());
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Yea, this is kinda hacky but it's fine. This is fixed in the version os pipelines that kestrel uses.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 3]
```diff
@@ -71,25 +71,25 @@ public Task StartAsync(Uri url, Channel<byte[], SendMessage> application)
...
             try
             {
                 while (true)
                 {
                     var result = await pipelineReader.ReadAsync();
                     var input = result.Buffer;
+                    if (result.IsCancelled || (input.IsEmpty && result.IsCompleted))
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Oh man... I'm going to have to fix this in corefxlab... `IsCancelled`

File: test/Microsoft.AspNetCore.SignalR.Client.Tests/ServerSentEventsTransportTests.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                        .Setup(s => s.CopyToAsync(It.IsAny<Stream>(), It.IsAny<int>(), It.IsAny<CancellationToken>()))
+                        .Returns(copyToAsyncTcs.Task);
+                    return new HttpResponseMessage { Content = new StreamContent(mockStream.Object) };
+                });
+
+            try
+            {
+                using (var httpClient = new HttpClient(mockHttpHandler.Object))
+                {
+                    var sse = new ServerSentEventsTransport(httpClient);
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.SignalR.Client.Tests/ServerSentEventsTransportTests.cs]: Remove this one, unused

**Comment by moozzyk** [comment]
:up::date:

**Pull request merged**